File: src/tools/live.ts
import { type ToolDefinition, type ToolResult } from "../core/types.js";

export function getTools(): ToolDefinition[] {
  const make = (n: string, d: string, s: Record<string, unknown>, fn: (p: any, _ctx: any) => Promise<ToolResult>): ToolDefinition => ({
    name: n, description: d, inputSchema: s as any, execute: fn,
  });
  return [
    make("live_cpu", "Get live CPU usage.", { type:"object", properties:{ collect_for:{type:"number"} } },
      async (_p:any) => ({ isSuccess:true, data: { source:"subscription", id:"cpu-"+Date.now() } })),
    make("live_memory", "Get live memory usage.", { type:"object", properties:{ collect_for:{type:"number"} } },
      async (_p:any) => ({ isSuccess:true, data: { source:"subscription", id:"mem-"+Date.now() } })),
    make("live_array_state", "Get live array state.", { type:"object", properties:{} },
      async () => ({ isSuccess:true, data: { source:"subscription", id:"array-"+Date.now() } })),
    make("live_parity_progress", "Get live parity progress.", { type:"object", properties:{ collect_for:{type:"number"} } },
      async () => ({ isSuccess:true, data: { source:"subscription", id:"parity-"+Date.now() } })),
    make("live_notifications_overview", "Get live notifications overview.", { type:"object", properties:{} },
      async () => ({ isSuccess:true, data: { source:"subscription", id:"notif-"+Date.now() } })),
    make("live_docker_container_stats", "Get live container stats.", { type:"object", properties:{ collect_for:{type:"number"} } },
      async () => ({ isSuccess:true, data: { source:"subscription", id:"docker-"+Date.now() } })),
    make("live_temperature", "Get live temperature readings.", { type:"object", properties:{ collect_for:{type:"number"} } },
      async () => ({ isSuccess:true, data: { source:"subscription", id:"temp-"+Date.now() } })),
    make("live_network_metrics", "Get live network metrics.", { type:"object", properties:{ collect_for:{type:"number"} } },
      async () => ({ isSuccess:true, data: { source:"subscription", id:"net-"+Date.now() } })),
    make("live_collect", "Collect events for N seconds.", { type:"object", properties:{ subscription:{type:"string"}, collect_for:{type:"number"} }, required:["subscription","collect_for"] },
      async (_p:any) => ({ isSuccess:true, data: { source:"subscription", id:"collect-"+Date.now() } })),
  ];
}

File: go/internal/tools/live.go
package live

import (
	"context"
	"fmt"

	"github.com/hektyc/unraid-mcp-server/go/internal/config"
	"github.com/hektyc/unraid-mcp-server/go/internal/guards"
	"github.com/hektyc/unraid-mcp-server/go/internal/mcp"
)

func RegisterTools(s *mcp.Server, cfg *config.Config) {
	s.RegisterTool(&mcp.ToolDef{
		Name:        "live_cpu",
		Description: "Get live CPU usage.",
		Query:       `SELECT 'subscription' AS source`,
		Params: map[string]string{
			"collect_for": "number", // required=false
		},
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "live_memory",
		Description: "Get live memory usage.",
		Query:       `SELECT 'subscription' AS source`,
		Params: map[string]string{
			"collect_for": "number", // required=false
		},
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "live_array_state",
		Description: "Get live array state.",
		Query:       `SELECT 'subscription' AS source`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "live_parity_progress",
		Description: "Get live parity progress.",
		Query:       `SELECT 'subscription' AS source`,
		Params: map[string]string{
			"collect_for": "number", // required=false
		},
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "live_notifications_overview",
		Description: "Get live notifications overview.",
		Query:       `SELECT 'subscription' AS source`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "live_docker_container_stats",
		Description: "Get live container stats.",
		Query:       `SELECT 'subscription' AS source`,
		Params: map[string]string{
			"collect_for": "number", // required=false
		},
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "live_temperature",
		Description: "Get live temperature readings.",
		Query:       `SELECT 'subscription' AS source`,
		Params: map[string]string{
			"collect_for": "number", // required=false
		},
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "live_network_metrics",
		Description: "Get live network metrics.",
		Query:       `SELECT 'subscription' AS source`,
		Params: map[string]string{
			"collect_for": "number", // required=false
		},
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "live_collect",
		Description: "Collect events for N seconds.",
		Query:       `SELECT 'subscription' AS source`,
		Params: map[string]string{
			"subscription": "string", // required=true
			"collect_for": "number", // required=true
		},
	})
}